### PR TITLE
G11n go service

### PR DESCRIPTION
**Pull request by vm-mounika**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  if [ x${TEST} = xcheck_header ]; then
+    git clone --branch=devops https://github.com/vmware/singleton.git devops
+    cp $TRAVIS_BUILD_DIR/devops/check_headers.py .
+    chmod +x check_headers.py
+    python3 ./check_headers.py -f "$(git diff --name-only --diff-filter=d $TRAVIS_COMMIT_RANGE)"
+  fi &&
+  if [ x${TEST} = xunit_test ]; then
+    make build
+    ls -ltr builds
+    make test
```
Comment (surite) [inline .travis.yml:32]: can you take a look at one failure in https://travis-ci.com/github/vm-mounika/singleton/jobs/498652991 line 1141 --- FAIL, is it a test case failure?

**Review by surite — CHANGES_REQUESTED** [review]
please setup your own sonar org and sonar token in your travis project  'vm-mounika / singleton', so sonar scan result will be in your own org, only the PR/branch result should be in singleton org.

File: .travis.yml (hunk 1 above)
Comment (Xiaochao8) [reply .travis.yml:32]: Yes, It its. This is caused by win/linux difference. https://github.com/vmware/singleton/pull/1077 is to fix this.

Comment (Xiaochao8) [reply .travis.yml:32]: Why unit test stage still passed?

Comment (vm-mounika) [reply .travis.yml:37]: @surite @Xiaochao8 I have moved build and test to separate steps.Now pipeline is failing with failing test ,Can you please check ?

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+language: go
+go:
+  - 1.13
```
[the remark this answers is not in the document]
Comment (vm-mounika) [reply .travis.yml]: @Xiaochao8 I have updated it.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+script: >
+  if [ x${TEST} = xcheck_header ]; then
+    git clone --branch=devops https://github.com/vmware/singleton.git devops
+    cp $TRAVIS_BUILD_DIR/devops/check_headers.py .
+    chmod +x check_headers.py
+    python3 ./check_headers.py -f "$(git diff --name-only --diff-filter=d $TRAVIS_COMMIT_RANGE)"
+  fi &&
+  if [ x${TEST} = xunit_test ]; then
+    make build
+    ls -ltr builds
```
[the remark this answers is not in the document]
Comment (vm-mounika) [reply .travis.yml]: @Xiaochao8 I moved it to build stage.
I removed ls.

File: .travis.yml [hunk 4]
```diff
@@ -26,10 +28,15 @@ script: >
     chmod +x check_headers.py
     python3 ./check_headers.py -f "$(git diff --name-only --diff-filter=d $TRAVIS_COMMIT_RANGE)"
   fi &&
-  if [ x${TEST} = xunit_test ]; then
+
+  if [ x${TEST} = xbuild ]; then
     make build
```
Comment (Xiaochao8) [inline .travis.yml:33]: Should check the result of command execution here?

Comment (vm-mounika) [reply .travis.yml:33]: Pipeline will check the execution status
Done. Your build exited with 0 or
Done. Your build exited with 1

Comment (Xiaochao8) [reply .travis.yml:33]: Could you check this PR https://github.com/vmware/singleton/pull/1077?
The CI is hanging.

File: .travis.yml (hunk 1 above)
Comment (Xiaochao8) [reply .travis.yml:37]: The pipeline is canceled. Do you know why?

Comment (vm-mounika) [reply .travis.yml:37]: @Xiaochao8 I was testing in my branch it is running other branch as well so canceled it.

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  fi &&
+
+  if [ x${TEST} = xcode_scan ]; then
+    SONARSCANNER='sonar-scanner-cli-4.0.0.1744-linux'
+    SONARSCANNERINNER='sonar-scanner-4.0.0.1744-linux'
+    wget https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/$SONARSCANNER.zip
+    unzip $SONARSCANNER.zip
+    rm -f $SONARSCANNER.zip
+    make test -json > report.json
+    make coverage
```
Comment (Xiaochao8) [inline .travis.yml:51]: Why do you run 'test' and 'coverage' in phase of 'code scan'?

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    SONARSCANNER='sonar-scanner-cli-4.0.0.1744-linux'
+    SONARSCANNERINNER='sonar-scanner-4.0.0.1744-linux'
+    wget https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/$SONARSCANNER.zip
+    unzip $SONARSCANNER.zip
+    rm -f $SONARSCANNER.zip
+    make test -json > report.json
+    make coverage 
+    git clone --branch=devops https://github.com/vmware/singleton.git devops
+    str=$(printf '=%.0s' {1..50})
+    echo $str Unit Test Start $str
```
Comment (Xiaochao8) [inline .travis.yml:54]: I see below messages in log. Please check.

str=$(printf '=%.0s' {1..50})
  echo $str Unit Test Start $str

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    cd $TRAVIS_BUILD_DIR/devops/sonar/ && python ./config_sonar_project.py -ProjectName=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -ProjectKey=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -OrgKey=$SONAR_ORG -QualityGateName=go-service-gate -QualityGateConditions="./quality_gates/go-service.json" -SonarToken=$SONAR_TOKEN
+    cd $TRAVIS_BUILD_DIR && ./$SONARSCANNERINNER/bin/sonar-scanner -Dsonar.sources=. -Dsonar.exclusions=**/testdata/**/*,**/$SONARSCANNERINNER/**/*,**/bindata/bindata.go,tests/**,api/v1/**,api/v2/swagger/**,assets/**,cmd/**,bindata/**,coverage.html -Dsonar.coverage.exclusions=**/testdata/**/*,**/$SONARSCANNERINNER/**/*,**/bindata/bindata.go,tests/**,api/v1/**,api/v2/swagger/**,assets/**,cmd/**,bindata/**,coverage.html -Dsonar.projectName=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -Dsonar.branch.name=master -Dsonar.projectKey=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -Dsonar.host.url=https://sonarcloud.io -Dsonar.organization=$SONAR_ORG -Dsonar.login=$SONAR_TOKEN -Dsonar.go.tests.reportPaths=./report.json -Dsonar.go.coverage.reportPaths=./cover.out
+    echo $str Unit Test End $str
```
Comment (Xiaochao8) [inline .travis.yml:57]: Same as previous one.

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    cd $TRAVIS_BUILD_DIR/server/publish && mv singleton-[0~9]*.jar i18n-service.jar
+    docker build -t singleton .
+    docker run -d -p 8090:8090 -p 8091:8091 --name singleton singleton
+    docker ps
+    sleep 15s
+    mkdir -p $TRAVIS_BUILD_DIR/l10n/bundles
+    cp -r $TRAVIS_BUILD_DIR/devops/autotest/client/go/l10n/bundles $TRAVIS_BUILD_DIR/l10n/
+    docker cp $TRAVIS_BUILD_DIR/l10n singleton:/
+    curl -X GET 'http://localhost:8091/i18n/api/v2/formatting/patterns?language=en&region=TW&scope=dates'        
+    cd $TRAVIS_BUILD_DIR/devops/autotest/client/go/goclienttest
```
Comment (Xiaochao8) [inline .travis.yml:84]: Why do you run go client tests against go service?

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    cd $TRAVIS_BUILD_DIR && ./$SONARSCANNERINNER/bin/sonar-scanner -Dsonar.sources=. -Dsonar.exclusions=**/testdata/**/*,**/$SONARSCANNERINNER/**/*,**/bindata/bindata.go,tests/**,api/v1/**,api/v2/swagger/**,assets/**,cmd/**,bindata/**,coverage.html -Dsonar.coverage.exclusions=**/testdata/**/*,**/$SONARSCANNERINNER/**/*,**/bindata/bindata.go,tests/**,api/v1/**,api/v2/swagger/**,assets/**,cmd/**,bindata/**,coverage.html -Dsonar.projectName=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -Dsonar.branch.name=master -Dsonar.projectKey=$SONAR_ORG-singleton-go-service-$TRAVIS_BRANCH -Dsonar.host.url=https://sonarcloud.io -Dsonar.organization=$SONAR_ORG -Dsonar.login=$SONAR_TOKEN -Dsonar.go.tests.reportPaths=./report.json -Dsonar.go.coverage.reportPaths=./cover.out
+    echo $str Unit Test End $str
+    git diff --name-only $TRAVIS_COMMIT_RANGE | grep -qvE '\.travis\.yml$' || {
```
[the remark this answers is not in the document]
Comment (vm-mounika) [reply .travis.yml:57]: This is for checking the quality gate.

File: .travis.yml (hunk 5 above)
Comment (vm-mounika) [reply .travis.yml]: Removed test, coverage is required.

File: .travis.yml (hunk 8 above)
Comment (vm-mounika) [reply .travis.yml]: Updated it.

**Review by surite — APPROVED** [review]

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    cd $TRAVIS_BUILD_DIR/devops/autotest/service/go/goservicetest
+    go get github.com/smartystreets/goconvey
+    echo "" >> go.mod
+    echo "replace github.com/vmware/singleton => $TRAVIS_BUILD_DIR" >> go.mod
+    echo "" >> go.mod
+    str=$(printf '=%.0s' {1..50})
+    echo $str Smoke Test Start $str
+    go test -v ./...
+    echo $str Smoke Test End $str
+  fi
```
Comment (Xiaochao8) [inline .travis.yml:92]: if [ x${TEST} = xsmoke_test ]; then
Download go service code and start service up, please refer to https://github.com/vmware/singleton/blob/master/.travis.yml
    git clone --branch=devops https://github.com/vmware/singleton.git devops
    cd $TRAVIS_BUILD_DIR/devops/autotest/service/goservice && gradle build
    str=$(printf '=%.0s' {1..50})
    echo $str Smoke Test Start $str
    Need to ask Yiyu how to run test. This is similiar to the Java version service. Below is an example.
    java -cp "target/*:resource/*" org.testng.TestNG testng.xml
    echo $str Smoke Test End $str
  fi

**Review by Xiaochao8 — CHANGES_REQUESTED** [review]
Please run CI passed first.

**Review by tigershi — APPROVED** [review]

**Pull request closed without merging**